### PR TITLE
[Validator] remove obsolete context and PropertyAccess code

### PR DESCRIPTION
**Pull request by Tobion**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php [hunk 1]
```diff
@@ -53,28 +43,8 @@ public function validate($value, Constraint $constraint)
...
-        } else {
-            $root = $this->context->getRoot();
-            $variables['value'] = $value;
-
-            if (is_object($root)) {
-                // Extract the object that the property belongs to from the object
-                // graph
-                $path = new PropertyPath($this->context->getPropertyPath());
-                $parentPath = $path->getParent();
-                $variables['this'] = $parentPath ? $this->getPropertyAccessor()->getValue($root, $parentPath) : $root;
```
Comment (Tobion) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: is this correct that all this is not needed anymore? To me it looks unreachable.

File: src/Symfony/Component/Validator/ValidatorBuilder.php [hunk 2]
```diff
@@ -90,11 +89,6 @@ class ValidatorBuilder implements ValidatorBuilderInterface
     private $translationDomain;
 
     /**
-     * @var PropertyAccessorInterface|null
-     */
-    private $propertyAccessor;
```
Comment (Tobion) [inline src/Symfony/Component/Validator/ValidatorBuilder.php]: it doesn't even have a setter anymore in 3.0. so it cannot be used.

File: src/Symfony/Component/Validator/ConstraintValidatorFactory.php [hunk 3]
```diff
@@ -42,7 +39,7 @@ public function getInstance(Constraint $constraint)
 
         if (!isset($this->validators[$className])) {
             $this->validators[$className] = 'validator.expression' === $className
```
Comment (Tobion) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: Do we need to keep this special name `validator.expression`? Or can I just remove that here and change the `validatedBy` implementation of `ExpressionValidator`. I don't see why there is this edge case for this.

Comment (Tobion) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: If it should be changed, what about the service definition in validator.xml, see above?

**Comment by Tobion** [comment]
@webmozart I need you to look over this please.

File: src/Symfony/Component/Validator/ConstraintValidatorFactory.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: I think we should remove the service (well we might have to keep it for BC reasons but deprecate it) and remove the special handling of the `ExpressionValidator` here as it is no longer needed.

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: Looks right to me (with the new API introduced in Symfony 2.5 we can now always access the object being validated through the context).

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php [hunk 4]
```diff
@@ -28,18 +24,12 @@
...
-
-    /**
      * @var ExpressionLanguage
      */
     private $expressionLanguage;
 
-    public function __construct(PropertyAccessorInterface $propertyAccessor = null, ExpressionLanguage $expressionLanguage = null)
+    public function __construct($propertyAccessor = null, ExpressionLanguage $expressionLanguage = null)
     {
-        $this->propertyAccessor = $propertyAccessor;
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: I would say let's deprecate the property accessor argument (like we did in Symfony 2.7 and 2.8 in several places).

**Comment by fabpot** [comment]
@Tobion I propose to merge this one as @webmozart does not reply anyway to pings.

**Comment by xabbuh** [comment]
@fabpot What do you think about deprecating the unneeded argument in Symfony 3.1 (see https://github.com/symfony/symfony/pull/17398#discussion_r49938630)?

**Comment by fabpot** [comment]
I'm not a big fan as this is a big BC break and makes supporting several versions difficult.

File: src/Symfony/Component/Validator/ConstraintValidatorFactory.php (hunk 3 above)
Comment (Tobion) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: well, the validator still has a dependency on expression language which could be injected as constructor argument. @fabpot do we not have a tag to add expression function providers to the expression language? this seems to be missing for the validator here. I guess it only support the default expression language so far.

File: src/Symfony/Component/Validator/ConstraintValidatorFactory.php [hunk 5]
```diff
@@ -26,11 +26,8 @@ class ConstraintValidatorFactory implements ConstraintValidatorFactoryInterface
 {
     protected $validators = array();
 
-    private $propertyAccessor;
-
-    public function __construct($propertyAccessor = null)
+    public function __construct()
```
Comment (webmozart) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: I suppose the empty constructor can now be removed?

Comment (webmozart) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: I guess you kept it for BC.

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php (hunk 1 above)
Comment (webmozart) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: :+1:

File: src/Symfony/Component/Validator/Constraints/ExpressionValidator.php (hunk 4 above)
Comment (webmozart) [inline src/Symfony/Component/Validator/Constraints/ExpressionValidator.php]: :+1:

**Comment by webmozart** [comment]
Looks good to me. @fabpot you think deprecating the argument is too much of a hassle even if we keep it until 4.0?

**Comment by fabpot** [comment]
@webmozart For ExpressionValidator, yes. It means that the constructor signature is going to be different for 3.4 and 4.0, which makes writing libraries/projects compatible with both versions more difficult (think Silex for instance).

File: src/Symfony/Component/Validator/ConstraintValidatorFactory.php (hunk 5 above)
Comment (Tobion) [inline src/Symfony/Component/Validator/ConstraintValidatorFactory.php]: yes

**Comment by Tobion** [comment]
I think it's good to merge. If we want to deprecate the ExpressionValidator argument or the `validator.expression` service, we would need to do it in the master branch anyway.

**Comment by xabbuh** [comment]
:+1: 

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @Tobion.

**Pull request closed without merging**